src/client/visualizations/base-visualization/base-visualization.tsx

```diff
@@ -79,33 +82,45 @@ export class BaseVisualization<S extends BaseVisualizationState> extends React.C
 
   componentWillReceiveProps(nextProps: VisualizationProps) {
     if (this.shouldFetchData(nextProps) && this.visualisationNotResized(nextProps)) {
-      this.fetchData(nextProps.essence, nextProps.timekeeper);
+      this.loadData(nextProps.essence, nextProps.timekeeper, !this.props.essence.equals(nextProps.essence));
     }
   }
 
-  protected fetchData(essence: Essence, timekeeper: Timekeeper): void {
-    this.lastQueryEssence = essence;
-    this.handleDatasetLoad(loading);
+  private loadData(essence: Essence, timekeeper: Timekeeper, showSpinner = true) {
+    if (showSpinner) this.handleDatasetLoad(loading);
+    this.fetchData(essence, timekeeper)
+      .then(dl => {
+        if (isLoading(dl)) return;
+        if (isError(dl)) {
+          this.handleDatasetLoad(dl);
+        }
+        if (isLoaded(dl)) {
+          this.handleDatasetLoad(dl, this.deriveDatasetState(this.props, dl.dataset));
+        }
+      });
+  }
 
-    this.debouncedCallExecutor(essence, timekeeper);
+  private fetchData(essence: Essence, timekeeper: Timekeeper): Promise<DatasetLoad> {
+    this.lastQueryEssence = essence;
+    return this.debouncedCallExecutor(essence, timekeeper);
   }
 
-  private callExecutor = (essence: Essence, timekeeper: Timekeeper) =>
+  private callExecutor = (essence: Essence, timekeeper: Timekeeper): Promise<DatasetLoad> =>
     essence.dataCube.executor(makeQuery(essence, timekeeper), { timezone: essence.timezone })
       .then((dataset: Dataset) => {
-          if (!this.wasUsedForLastQuery(essence)) return;
-          this.handleDatasetLoad(loaded(dataset), this.deriveDatasetState(this.props, dataset));
+          if (!this.wasUsedForLastQuery(essence)) return loading;
```

Comment:
For old promises we return loading as "silent fail". Maybe we should be more explicit? Pass null there and handle null in `then`